lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -2461,56 +2530,67 @@ internal class EvaluatingCompiler(
+                    (escapeExpr as? PartiqlAst.Expr.Lit)?.value?.toIonValue(valueFactory.ion)
+                        ?.let { valueFactory.newFromIonValue(it) })
 
                 // If valueExpr is also a literal then we can evaluate this at compile time and return a constant.
-                if (valueExpr is Literal) {
-                    val resultValue = runPatternParts(valueFactory.newFromIonValue(valueExpr.ionValue), patternParts)
-                    return thunkFactory.thunkEnv(operatorMetas) { resultValue }
+                if (valueExpr is PartiqlAst.Expr.Lit) {
+                    val resultValue = runPatternParts(
+                        valueFactory.newFromIonValue(valueExpr.value.toIonValue(valueFactory.ion)),
+                        patternParts
+                    )
+                    return thunkFactory.thunkEnv(metas) { resultValue }
                 } else {
-                    thunkFactory.thunkEnvOperands(operatorMetas, valueThunk) { _, value ->
+                    thunkFactory.thunkEnvOperands(metas, valueThunk) { _, value ->
                         runPatternParts(value, patternParts)
                     }
                 }
             }
             else -> {
-                val patternThunk = argThunks[1]
-                when (argThunks.size) {
-                    2 -> {
+                val patternThunk = compileAstExpr(patternExpr)
+                when (escapeExpr) {
+                    null -> {
                         //thunk that re-compiles the DFA every evaluation without a custom escape sequence
-                        thunkFactory.thunkEnvOperands(operatorMetas, valueThunk, patternThunk) { _, value, pattern ->
+                        thunkFactory.thunkEnvOperands(metas, valueThunk, patternThunk) { _, value, pattern ->
                             val pps = getPatternParts(pattern, null)
                             runPatternParts(value, pps)
                         }
                     }
                     else -> {
                         //thunk that re-compiles the pattern every evaluation but *with* a custom escape sequence
-                        val escapeThunk = argThunks[2]
-                        thunkFactory.thunkEnvOperands(operatorMetas, valueThunk, patternThunk, escapeThunk) { _, value, pattern, escape ->
+                        val escapeThunk = compileAstExpr(escapeExpr)
+                        thunkFactory.thunkEnvOperands(
+                            metas,
+                            valueThunk,
+                            patternThunk,
+                            escapeThunk
+                        ) { _, value, pattern, escape ->
                             val pps = getPatternParts(pattern, escape)
                             runPatternParts(value, pps)
                         }
                     }
                 }
             }
         }
+
+        return resolveIntConstraint(computeThunk, metas)
```

Comment:
`LIKE` returns a boolean, so this Int check can be removed.

Reply:
Will remove it in the next commit